File: 2021/06/09/main.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        dp = -nums[-1]
        pq = [(dp, len(nums) - 1)]
        for i in range(len(nums) - 2, -1, -1):
            while pq[0][1] > i + k:
                heapq.heappop(pq)
            dp = pq[0][0] - nums[i]
            heapq.heappush(pq, (dp, i))
        return -dp

File: 2021/06/09/main.go
package main

func maxResult(nums []int, k int) int {
	n := len(nums)

	dp := make([]int, n)
	dp[0] = nums[0]

	q := make([]int, 0)
	q = append(q, 0)

	for i := 1; i < n; i++ {

		for len(q) > 0 && q[0] < i-k {
			q = q[1:]
		}

		dp[i] = dp[q[0]] + nums[i]

		for len(q) > 0 && dp[i] > dp[q[len(q)-1]] {
			q = q[:len(q)-1]
		}

		q = append(q, i)
	}

	return dp[n-1]
}
